### PR TITLE
Setup dependency injection for mocks

### PR DESCRIPTION
**Pull request by ThomasMPont**
## Motivation

We need to write end-to-end tests before this week's milestone. To achieve this, we must mock all internet or account related backend logic. This is done using dependency injection, where a mock class replaces default implementations of backend logic in a screen's parameters.

Clearing up space for dependency injection will also enable more in depth integration tests. Helping improve coverage for those difficult connectivity related classes.

## Changes

This PR only aims to make space for future changes. It should not affect the app's functionality in any way.
I would still recommend reviewers to thouroughly check the app's logic, as I might have inadvertently broken something.

`MainActivity's` `SignifyAppPreview` now takes a `DependencyProvider` parameter. This dependency provider is a singleton object with methods that provide the app with it's backend logic classes.

By default, the app uses `AppDependencyProvider` which provides the expeted classes, such as `UserRepositoryFirestore`. In the future, end-to-end tests will use `MockDependencyProvider` which will replace backend logic with simpler and offline mock versions of these classes. These mock classes have not yet been implemented.

This change allows developers to freely define new dependency providers, that alter the backend logic of the app for testing or demo purposes.

## Future changes

The current implementation of the login screen uses Google authentication through Firebase directly. This login logic must be moved behind an `interface class` so that it can be mocked. Mocking the login screen is essential to writing end-to-end tests.

At the moment, `MockDependencyProvider` only injects dummy classes into the app. It should be augmented with actual mocks of backend logic to allow for realistic system tests or integration tests.

Finally, we obviously need to write end-to-end tests that use this mock dependency injection.

### CODE REVIEW

File: app/src/androidTest/java/com/github/se/signify/ui/screens/MainActivityTest.kt [hunk 1]
```diff
@@ -33,15 +33,14 @@ class MainActivityTest {
...
     val context = mock(Context::class.java)
 
     composeTestRule.setContent {
-      FriendsListScreen(navigationActions, userViewModel)
-      SettingsScreen(navigationActions, userViewModel)
-      ProfileScreen(navigationActions, userViewModel)
+      FriendsListScreen(navigationActions, userRepository)
+      SettingsScreen(navigationActions, userRepository)
+      ProfileScreen(navigationActions, userRepository)
       // Set the content with the mocked context
```
Comment (NBCCF025) [inline app/src/androidTest/java/com/github/se/signify/ui/screens/MainActivityTest.kt:42]: Is it necessary to add the `FriendsListScreen()`, `SettingsScreen()` and `ProfileScreen()` here as they are already called with `SignifyAppPreview()`.

**Review by nada-ezziti — APPROVED** [review]
### Pull Request Review: Feedback on Dependency Injection Implementation

#### **Strengths**
1. **Solid Foundation for Testing**: Introducing `DependencyProvider` and its implementations is an excellent step toward improving testability. The design aligns with best practices in dependency injection, which will streamline the creation of robust end-to-end and integration tests.
   
2. **Minimal Disruption**: The changes are well-scoped and intentionally non-intrusive. By maintaining existing functionality and introducing mock capability incrementally, this PR reduces the risk of regressions.

3. **Flexibility for Future Development**: Allowing developers to define custom dependency providers enables greater flexibility for testing, demoing, and even prototyping. This is a smart move to future-proof the codebase.

4. **Commit Organization**: The commits are clean and logical, making it easier to review and understand the progression of changes.

---

#### **Remarks**

1. **Testing Strategy**: Although this PR is primarily preparatory, it would be helpful to include a minimal test case demonstrating how `MockDependencyProvider` could be used. This serves as a proof of concept and ensures the current setup supports the intended use case.


2. **Risk of Inadvertent Breakage**: The note about potential breakages is valid, especially given the changes to `MainActivity` and ViewModel parameters. It’s critical to have regression tests covering all screens that interact with `DependencyProvider`. Adding a checklist for reviewers to test key flows (e.g., login, profile, challenges) could be helpful.
---

#### **Overall Feedback**
This PR lays a strong foundation for end-to-end testing and improves the app’s architecture without introducing unnecessary complexity. Great job!

File: app/src/main/java/com/github/se/signify/ui/screens/profile/FriendsListScreen.kt [hunk 2]
```diff
@@ -59,7 +60,8 @@ val currentUserId = FirebaseAuth.getInstance().currentUser?.email?.split("@")?.g
 @Composable
 fun FriendsListScreen(
     navigationActions: NavigationActions,
-    userViewModel: UserViewModel = viewModel(factory = UserViewModel.Factory)
+    userRepository: UserRepository,
+    userViewModel: UserViewModel = viewModel(factory = UserViewModel.factory(userRepository))
```
Comment (NBCCF025) [inline app/src/main/java/com/github/se/signify/ui/screens/profile/FriendsListScreen.kt:64]: The userViewModel could be a val outside the parameter list of `FriendsListScreen()` to keep coherence with other screen implementation.

**Review by NBCCF025 — APPROVED** [review]
This PR introduces the initial setup for end-to-end test support through dependency injection, allowing for mock implementations of backend logic. This is a valuable foundation to facilitate testing for complex network and account-related classes, ultimately improving test coverage and ensuring milestone stability.

Excellent work, I approve this PR.

File: app/src/androidTest/java/com/github/se/signify/ui/screens/MainActivityTest.kt (hunk 1 above)
Comment (ThomasMPont) [reply app/src/androidTest/java/com/github/se/signify/ui/screens/MainActivityTest.kt:42]: I haven't touched this test. The function calls have only been updated with their new parameters.

File: app/src/main/java/com/github/se/signify/ui/screens/profile/FriendsListScreen.kt (hunk 2 above)
Comment (ThomasMPont) [reply app/src/main/java/com/github/se/signify/ui/screens/profile/FriendsListScreen.kt:64]: I had originally done that. Turns out some weird tests in `FriendsListScreenTest` are passing over a `ViewModel` they have a reference to into the screen to test the viewmodel's logic. I don't think these tests should even exist, but left looking into it for a later task.

I should've mentioned this in my description. Good point.

**Comment by ThomasMPont** [comment]
> #### **Remarks**
> 
> 1. **Testing Strategy**: Although this PR is primarily preparatory, it would be helpful to include a minimal test case demonstrating how `MockDependencyProvider` could be used. This serves as a proof of concept and ensures the current setup supports the intended use case.
> 
> 
> 2. **Risk of Inadvertent Breakage**: The note about potential breakages is valid, especially given the changes to `MainActivity` and ViewModel parameters. It’s critical to have regression tests covering all screens that interact with `DependencyProvider`. Adding a checklist for reviewers to test key flows (e.g., login, profile, challenges) could be helpful.

1. Since no mocks have been created for the `MockDependencyProvider`, and it isn't used anywhere yet, I couldn't find a pertinent way to test it. You're right I probably should have come up with some basic test for it though.

2. I like the idea of a checklist. I suggest we write a list of all the app's features in the app's Wiki, that would be gone over before major changes or milestones.

**Pull request merged**